ngraph/core/reference/include/ngraph/runtime/reference/gather.hpp

```diff
@@ -27,149 +30,386 @@ namespace ngraph
+                        T* out_outer = &out[out_outer_transform.index(*out_outer_coord_iter)];
+
+                        auto out_inner_coord_iter = out_inner_transform.begin();
+                        for (const Coordinate& indices_outer_coord : indices_outer_transform)
+                        {
+                            if (out_inner_coord_iter == out_inner_transform.end())
+                                break;
+                            const U* indices_prime =
+                                &indices[indices_outer_transform.index(indices_outer_coord)];
+                            T* out_prime =
+                                &out_outer[out_inner_transform.index(*out_inner_coord_iter)];
+                            gather_nd<T, U>(params_prime,
+                                            indices_prime,
+                                            out_prime,
+                                            params_prime_shape,
+                                            indices_prime_shape,
+                                            out_prime_shape);
+                            ++out_inner_coord_iter;
+                        }
+                        ++out_outer_coord_iter;
+                    }
                 }
-                for (size_t i = indices_ndim; i < out_inner_ndim; i++)
+            } // namespace v0
+
+            namespace v1
+            {
+                namespace
                 {
-                    out_inner_end_corner[i] = 1;
-                }
-                const Strides out_inner_strides(out_inner_ndim, 1);
-                AxisVector out_inner_axis_order(out_inner_ndim);
-                std::iota(out_inner_axis_order.begin(), out_inner_axis_order.end(), 0);
-                const CoordinateTransform out_inner_transform(out_inner_shape,
-                                                              out_inner_start_corner,
-                                                              out_inner_end_corner,
-                                                              out_inner_strides,
-                                                              out_inner_axis_order);
-
-                auto out_outer_coord_iter = out_outer_transform.begin();
-                for (const Coordinate& params_outer_coord : params_outer_transform)
+                    template <bool check>
+                    using Required = typename std::enable_if<check, bool>::type;
+
+                    template <typename It>
+                    struct IsRandomAccessIt
+                    {
+                        static constexpr bool value =
+                            std::is_same<typename It::iterator_category,
+                                         std::random_access_iterator_tag>::value;
+                    };
+
+                    /// @brief Span should mimic std::span
+                    template <typename Element>
+                    class Span
```

Comment:
The same class is implemented in gather_nd operation. The logic is quite similar and I propose to implement common class for span and use it in both places. Pointers and iterators have the same behavior, and the span class can be implemented for both of them.

Reply:
You have absolutely right. This is more or les the same class. At first it was copy-paste from gather_nd with some extensions. 

@vladislav-volkov , do we have good place for such general tool (it could be use widely)?

Reply:
@pelszkow At the current moment we don't have  a place for such tools.
I propose to create a `utils` or `details` folder and put the `Span` class in this folder.
In should allow to detach the utils from reference implementations.

Reply:
Done